File: Step 07 Recursion/Step 7.2 Subsequence Pattern/Count all subsequences with sum K/Count all subsequences with sum K.java


class Solution{
	public int perfectSum(int arr[], int n, int sum) {
//         if (n == 1) return (arr[0] == target) ? 1 : 0;
//         // Arrays.sort(arr);
//         HashMap<Integer, Integer> preSum = new HashMap<>();
//         int count = 0;
//         int left = 0;
//         int right = 0;
//         int currSum = arr[0];
//         preSum.put(arr[0], 1); // Initialize with 1 way to get currSum
//         int MOD = 1000000007; // Modulo value
//         while ((left <= right) && right < n) {
//             if (preSum.containsKey(target - currSum)) {
//                 count = (count + preSum.get(target - currSum)) % MOD;
//             }
    
//             if (currSum < target) {
//                 right++;
//                 if (right < n) {
//                     currSum = (currSum + arr[right]) % MOD;
//                 }
//             } else if (currSum >= target) {
//                 currSum = (currSum - arr[left]) % MOD;
//                 left++;
//             }
//             // Update the count of ways to get currSum
//             preSum.put(currSum, (preSum.getOrDefault(currSum, 0) + 1) % MOD);
//         }
    
//         return count;
int MOD = (int)1e9+7;

        // Initialize a 1D array to store the count of subsets
        int[] dp = new int[sum + 1];

        // There is one way to make sum = 0, which is by not including any element
        dp[0] = 1;

        // Fill the dp array using a bottom-up approach
        for (int i = 0; i < n; i++) {
            for (int j = sum; j >= arr[i]; j--) {
                // If the current element can be included in the subset
                // (i.e., it is less than or equal to the current sum),
                // update dp[j] by adding dp[j - arr[i]]
                dp[j] = (dp[j] + dp[j - arr[i]]) % MOD;
            }
        }

        // The result is stored in dp[sum]
        return dp[sum];
}
}

File: Step 07 Recursion/Step 7.2 Subsequence Pattern/Count all subsequences with sum K/Count all subsequences with sum K.cpp
// DP APPROACH

class Solution {
public:
    int MOD = 1e9 + 7;

    int perfectSum(int arr[], int n, int sum) {
        vector<vector<int>> dp(n + 1, vector<int>(sum + 1, 0));
        
        // There's exactly one way to achieve a sum of 0: using no elements
        dp[0][0] = 1;
        
        for (int i = 1; i <= n; ++i) {
            for (int j = 0; j <= sum; ++j) {
                // Case 1: Exclude the current element
                dp[i][j] = dp[i - 1][j];
                
                // Case 2: Include the current element (if it's not greater than the target sum)
                if (j >= arr[i - 1]) {
                    dp[i][j] = (dp[i][j] + dp[i - 1][j - arr[i - 1]]) % MOD;
                }
            }
        }
        
        return dp[n][sum];
    }
};


// RECURSIVE APPROACH

// class Solution{
// 	public:
// 	int MOD = 1e9 + 7;
	// int countDistinctSubsequences(int arr[], int n, int sum, int currSum, int index) {
    //     if (index == n) {
    //         if (currSum == sum) {
    //             return 1;
    //         }
    //         return 0;
    //     }

    //     // Case 1: Exclude the current character and move to the next
    //     int exclude = countDistinctSubsequences(arr, n, sum, currSum, index + 1);

    //     // Case 2: Include the current character and move to the next
    //     int include = countDistinctSubsequences(arr, n, sum, currSum + arr[index], index + 1);

    //     return (exclude + include) % MOD;
    // }

// 	int perfectSum(int arr[], int n, int sum)
// 	{
//         // Your code goes here
//         return countDistinctSubsequences(arr, n, sum, 0, 0);
// 	}
// };